Modifation to allow mixed element types in matrix operations

awbsmith opened the pull request:
This should enable things like automatic differentiation:

```Julia
using FixedSizeArrays
using ForwardDiff

R = rand(Mat{2,2,Float64})
X = Vec(ForwardDiff.GradientNumber(0.0, (1.0, 0.0)), ForwardDiff.GradientNumber(1.0, (0.0, 1.0)))

Y = R * X
```

As per [issue 93](https://github.com/SimonDanisch/FixedSizeArrays.jl/issues/93)

mschauer commented on src/ops.jl:
```diff
@@ -115,21 +115,39 @@ end
-@inline bilindot{T}(a::NTuple{3,T}, b::NTuple{3,T}) = @inbounds return (a[1]*b[1] + a[2]*b[2] + a[3]*b[3])
-@inline bilindot{T}(a::NTuple{4,T}, b::NTuple{4,T}) = @inbounds return (a[1]*b[1] + a[2]*b[2] + a[3]*b[3]+a[4]*b[4])
+
+immutable BilinearDotFunctor <: Func{2} end
```
Double line

mschauer commented:
@c42f Will this be superseded by your work?

mschauer commented on src/ops.jl:
```diff
@@ -115,21 +115,39 @@ end
+
+immutable BilinearDotFunctor <: Func{2} end
+call(::BilinearDotFunctor, a, b) = a*b
+@inline bilindot{T <: Union{FixedArray, Tuple}}(a::T, b::T) = sum(map(DotFunctor(), a, b))
```
Note #107

c42f commented on src/ops.jl:
```diff
@@ -115,21 +115,39 @@ end
 
 #cross{T}(a::FixedVector{2, T}, b::FixedVector{2, T}) = a[1]*b[2]-a[2]*b[1] # not really used!?
-@inline cross{T<:Number}(a::FixedVector{3, T}, b::FixedVector{3, T}) = @inbounds return typeof(a)(
+@inline cross{T1 <: Number, T2 <: Number}(a::FixedVector{3, T1}, b::FixedVector{3, T2}) = @inbounds return Vec(
```
This shouldn't really return a `Vec`, as `a` and `b` could be any user-defined type.

What type should it return?  I guess `similar(typeof(a), promote_type(T1,T2))` would be somewhat appropriate, though not really guaranteed to make sense for all user defined FSAs.  (And what about `b` anyway?  Well `Base` prefers the type of the left hand argument when deducing the output container to pass to `similar` for things like `+`, so I guess that'll do for now.)

mschauer replied:
Maybe a bit too much, but as the natural unit of the cross product vector is <s>volume</s> area and not length, something like http://docs.julialang.org/en/release-0.4/devdocs/promote-op/ or newer http://docs.julialang.org/en/latest/devdocs/promote-op/ might be possible. The reductions in base use this advance promotion, right?

c42f replied:
True, `cross` should return something with volume units in principle.

`promote_op` is one solution, and I'm probably going to use that elsewhere but it's a bit dissatisfyingly manual.  The other would be a hypothetical `similar`-like function which constructs FSAs out of provided tuples.  This way, the element type can be inferred from the result type of expressions such as `a[2]*b[3]-a[3]*b[2]` - like is being done here with the `Vec` constructor - but without hardcoding the output type.

``` julia
# Possible (untested!) default, not sure about the name...
construct_similar{F<:FixedArray}(a::F, elements::Tuple) = basetype(F)(promote_type(elements...))
```

andyferris replied:
Do we mean _area_ units? The cross product gives the area of a parallelogram. Take a further dot product with a length vector to get a volume.

mschauer replied:
Yes, directed area, not directed volume! Thank you.

c42f replied:
Eh, I copied that one right over without even noticing, quite right it's the area.

c42f commented:
@mschauer I don't think so.  I'm currently mostly messing with elementwise operations for which `map` is the right thing, but the matrix multiply stuff here seems largely independent.

awbsmith commented:
Hi all, I've:

1) made the line space change
2) changed the cross product to return `similar(typeof(a), promote_type(T1,T2))` as per @c42f https://github.com/c42f's suggestion. I don't think it's practical to know the return type, e.g. if its `cross(displacement, force)` you'd need a type defined for torque.  And returning the type of `a` will probably be less annoying than returning a generic Vec.  The good news is the change to the upstream master have made similar type safe.
3) rebased for the changes to the upstream master (and made a mess of it requiring a git history "fix", so sorry about the comment links)

Cheers,
Andrew

On 12 May 2016 at 16:34, Coveralls notifications@github.com wrote:

> [image: Coverage Status] https://coveralls.io/builds/6148606
> 
> Coverage decreased (-1.7%) to 91.362% when pulling _200bab0
> https://github.com/SimonDanisch/FixedSizeArrays.jl/commit/200bab0b6d5912fe74fb9c2269ff03443bb4d011
> on awbsmith:MatrixOpsElementPromotion_ into _a501d25
> https://github.com/SimonDanisch/FixedSizeArrays.jl/commit/a501d252164593e52892fbc11aedef83933e1579
> on SimonDanisch:master_.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/SimonDanisch/FixedSizeArrays.jl/pull/98#issuecomment-218672120

c42f commented:
Thanks Andrew.  This is now superseded by #119 which also includes the changes here.
